### ISSUE TITLE
Cache misses when building two NextJS apps using .env files

### ISSUE THREAD

**Issue by johnschult** [vercel/turborepo#2065]
I am running the `build` and `lint` tasks and filtering only the two Next JS apps, I get  full turbo one time, then cache miss the next, and then maybe I will get three full turbos, then another cache miss. It seems random, although I doubt it is 😄 

Something else worth noting (or not) is that I am running `build` for NextJS with the command `next build --no-lint` (because I want more control via turbo for when `lint` is executed), and my `lint` command is `next lint --no-cache`. I added in `--no-cache` because I was seeing empty results (using TIMING=1) when `next lint` was cached internally (not via turbo).

Both of the apps have no `pnpm` workspace dependencies other than:

```json
"devDependencies": {
   "tsconfig": "workspace:*",
}
```

_Originally posted by @johnschult in https://github.com/vercel/turborepo/issues/2029#issuecomment-1256461505_

**Comment by johnschult** [comment]
After messing a bit with a clean `my-turborepo` app, I am seeing that simply adding an `.env` file with `FOO=foo` to one of the apps causes a cache miss. I am thinking that my problem stems from that in my own app. It is possible that I am not completely understanding how environment variables affect the cache hash. I thought they would not matter unless you declared them in the pipeline.

We use `.env` because our apps are using NextAuth and we need to set certain environment variables in order for NextAuth to function correctly, however, although both apps use the same variables, they do not use the same values.

https://next-auth.js.org/configuration/options

What am I missing here? What might be a better practice?

**Comment by johnschult** [comment]
I am going to try to use a single `.envrc` file in the monorepo root and set specific app environment variables like:

```sh
APP_A_NEXTAUTH_SECRET=foo
```
 and then read them during runtime and set the options instead of relying on the default NextAuth ones. Maybe that will resolve this. More news at 11...

**Comment by johnschult** [comment]
The above adventure made no difference, still getting very inconsistent caching, mostly not... even when I filter and only build on app.

**Comment by gsoltis** [comment]
`.env` files tricky, depending on how you're using them. If they are checked in to git, they'll be included in the hash by default. If they are git ignored'd, but you want to depend on them, add them to `dependsOn`. The environment variables are added after `turbo` starts executing your task, so we don't directly have visibility into them.

**Comment by gsoltis** [comment]
Can you try w/ `1.5.3` when you get a chance?

**Comment by gsoltis** [comment]
I think this is resolved w/ `1.5.4`. If not, let me know and I'll reopen
